Enable Azure Workload Identity to authorize against RabbitMQ manageme…

kratkyzobak opened the pull request:
<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md
-->

This PR provides support for Azure AD Workload Identity authorization against RabbitMQ HTTP API in RabbitMQ scaler. This utilizes generic OIDC support introduced in RabbitMQ 3.11

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [X] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [X] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)* - not applicable
- [X] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)* - https://github.com/kedacore/keda-docs/pull/1156
- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

<!--
  Make sure to link the related issue for this change
  If it requires multiple PRs and/or a PR on another repo as well, please use "Relates to" instead.
-->
Fixes #4716

<!--
  Make sure to link the related PRs for changes such as documentation & Helm charts
-->
Relates to #

kratkyzobak commented:
I'm currently not sure how to provide e2e test for this. I have checked, that workload identity are not checked even in most of azure scalers. To provide e2e test, there would need to be Azure AD app registration in test environment. I can provide with details about configuring this app registration in test tenant if needed.

zroubalik commented:
> I'm currently not sure how to provide e2e test for this. I have checked, that workload identity are not checked even in most of azure scalers. To provide e2e test, there would need to be Azure AD app registration in test environment. I can provide with details about configuring this app registration in test tenant if needed.

@JorTurFer can help with this one, most likely 😄

JorTurFer commented:
> I have checked, that workload identity are not checked even in most of azure scalers. 

What do you mean? I believe that all the Azure scalers are tested without any pod identity, with AAD pod identity and also with workload identity.
During the e2e startup, workload identity (and others) is deployed.

What would you need for adding the e2e test? I guess that you need an Azure App registration which exposes an api, and granting permissions over that api, am I right?

We have a repository for managing e2e test infrastructure using terraform (the repo is testing-infrastructure), if you open a PR there adding whatever you need, it will be available for e2e tests. We can help you too if you explain us what do you need exactly

kratkyzobak commented:
> testing-infrastructure

kratkyzobak commented:
> What would you need for adding the e2e test? I guess that you need an Azure App registration which exposes an api, and granting permissions over that api, am I right?
> 
> We have a repository for managing e2e test infrastructure using terraform (the repo is testing-infrastructure), if you open a PR there adding whatever you need, it will be available for e2e tests. We can help you too if you explain us what do you need exactly

Yeah, I would need application registration in Azure AD to use as OAuth scope. I will prepare PR into testing-infrastructure with proper  `azuread_application` (would be almost same as [rabbit-api resource mentoined here](https://github.com/rabbitmq/rabbitmq-server/issues/7861#issuecomment-1503069506)). Since there is no `azuread_*` resource within whole testing-infrastructure project, but `azuread` provider is registered - will SP running github action has right to make app registration, create enterprise app (azuread_service_principal) and assign role of this app to some of managed identity (azuread_app_role_assignment)?

JorTurFer commented:
> Since there is no `azuread_*` resource within whole testing-infrastructure project, but `azuread` provider is registered - will SP running github action have right to make app registration, create enterprise app (azuread_service_principal) and assign role of this app to some of managed identity (azuread_app_role_assignment)?

I think so, but if it couldn't, we can escalate the SP permissions

kratkyzobak commented:
I have added e2e tests, but it needs to wait until resolve https://github.com/kedacore/testing-infrastructure/pull/114#issuecomment-1593391936. With infrastructure locally created using terraform-infrastructure, added E2E tests are ok.

OT: Are there any logs of previous e2e tests runs available publicly? I cannot find any and I cannot understand how would current rabbitmq tests work. There were missing imports (so tests wont compile), wrong `.env` file was loaded (probably not problem in Actions runner). Or am I missing something?

JorTurFer commented:
/run-e2e rabbit

JorTurFer commented:
Hi,
You can see the logs clicking in the comment:
![image](https://github.com/kedacore/keda/assets/36899226/b3fe07ce-857a-48d3-b32b-0fbddb6a759a)

JorTurFer commented:
Hi @kratkyzobak ,
DCO check is failing and that's because there is some commit not signed. Could you fix it?
https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#i-didnt-sign-my-commit-now-what

kratkyzobak commented:
@JorTurFer fixed. Sorry for that. I used rebase button in Github, but I have different email locally than in Github account.

zroubalik commented:
/run-e2e rabbit

zroubalik commented on CHANGELOG.md:
```diff
@@ -62,6 +62,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **Pulsar Scaler**: Improve error messages for unsuccessful connections ([#4563](https://github.com/kedacore/keda/issues/4563))
 - **Security**: Enable secret scanning in GitHub repo
 - **RabbitMQ Scaler**: Add support for `unsafeSsl` in trigger metadata ([#4448](https://github.com/kedacore/keda/issues/4448))
+- **RabbitMQ Scaler**: Add support for `workloadIdentityResource` and utilize AzureAD Workload Identity for HTTP authorization
```
Could you please open an issue for this and reference it here? Thanks!

zroubalik commented:
/run-e2e rabbit

zroubalik reviewed (approved):
LGTM

Díky!
